Time zone

antoine-de opened the pull request:
## PR for time zone handling

The main idea is to convert ALL times to UTC  in the import process, handle only UTC in kraken (the computation part) and change it to local time only at the end of the process in Jörmungandr (the middleware) using the added time zone info in each stop areas.

The UTC conversion difficult part is day saving time (DST) handling.  
For example a stoptime at 7am valid over the year in a paris (Europe/Paris timezone) will be 6AM UTC in winter but 5AM UTC in summer.

Thus all validity_pattern (gtfs calendar), vehicule_journey (gtfs trips) and stoptime have to be split over the DST period.
## Note

despite all its flaws we use boost::date_time for timzeone handling.
for the moment the boost csv is hardlinked in the project, but it will change (maybe by patching boost for it to read the  tzdb).

antoine-de committed:
Ed: reading time zone in GTFS

for the moment the boost tz db (why didn't they use the olson db?..) is
loaded in the repository for convenience.

real work is starting now, with the utc conversion of all the stop times

antoine-de committed:
Ed: reading time zone in GTFS

for the moment the boost tz db (why didn't they use the olson db?..) is
loaded in the repository for convenience.

real work is starting now, with the utc conversion of all the stop times

antoine-de committed:
Ed: no need to init the vp to 0 it is done by default

antoine-de committed:
Ed: no need to init the vp to 0 it is done by default

antoine-de committed:
Ed: split on dst

antoine-de committed:
Ed: split on dst

antoine-de committed:
Ed: tz handling for GTFS

fusio still need to be done

antoine-de committed:
Ed: tz handling for GTFS

fusio still need to be done

antoine-de committed:
Ed: cout cleaning before PR

antoine-de committed:
Ed: cout cleaning before PR

antoine-de committed:
Ed: conversion frequencies stop times to UTC

antoine-de committed:
Ed: conversion frequencies stop times to UTC

antoine-de committed:
Ed: fusio parser to utc

antoine-de committed:
Ed: default timezone in fusio

antoine-de committed:
Ed: default timezone in fusio

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -214,35 +214,35 @@ void StopTimeFusioHandler::init(Data& data) {
+            uint16_t local_traffic_zone =  boost::lexical_cast<uint16_t>(row[itl_c]);
+            if (local_traffic_zone > 0)
+                stop_time->local_traffic_zone = local_traffic_zone;
+            else
```
ça ne sert à rien

l-vincent-l commented on source/ed/connectors/gtfs_parser.cpp:
```diff
@@ -46,6 +46,113 @@ namespace ed{ namespace connectors {
+    return res;
+}
+
+std::vector<period_with_utc_shift> split_over_dst(const boost::gregorian::date_period& validity_period, const boost::local_time::time_zone_ptr& tz) {
```
ça serait pas mal de passer une ligne

antoine-de committed:
ED: add stop area's timezone to database

antoine-de committed:
All: change protobuff dates from string to int

date_times are now posix time stamp (nb seconds since 1970/01/01).

when hour are given, they are express as number of seconds since
midnight)
when date are given they are express as a posix timestamp to midnight

antoine-de committed:
All: change protobuff dates from string to int

date_times are now posix time stamp (nb seconds since 1970/01/01).

when hour are given, they are express as number of seconds since
midnight)
when date are given they are express as a posix timestamp to midnight

antoine-de committed:
Jormungandr: convert utc to timezone in output

All datetimes are in utc in the core, and they are converted in local
(depending on the stop area) in output

antoine-de committed:
Jormungandr: convert utc to timezone in output

All datetimes are in utc in the core, and they are converted in local
(depending on the stop area) in output

antoine-de committed:
All: handle posix time from query to endpoint

antoine-de committed:
All: handle posix time from query to endpoint

antoine-de committed:
Jormungandr: add a debug mode

a some misc corrections

antoine-de committed:
Jormungandr: add a debug mode

a some misc corrections

antoine-de committed:
Jormungandr: change timezone output handling

Now ALL date times are given in the dataset timezone not anymore in the
nearest stop areas time zone.

antoine-de committed:
Jormungandr: change timezone output handling

Now ALL date times are given in the dataset timezone not anymore in the
nearest stop areas time zone.

antoine-de committed:
Jormungandr: add tests on journeys

antoine-de committed:
Jormungandr: add tests on journeys

antoine-de committed:
Jormungandr: squeleton for split date time

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -65,6 +65,31 @@ void AgencyFusioHandler::handle_line(Data& data, const csv_row& row, bool) {
+
+    std::string timezone_name = row[time_zone_c];
+    if (! is_first_line) {
+        //we created a default agency, with a default time zone, but this will be overidden is we read at least one agency
```
if pas is

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -65,6 +65,31 @@ void AgencyFusioHandler::handle_line(Data& data, const csv_row& row, bool) {
+    }
+
+    if (timezone_name.empty()) {
+        throw navitia::exception("Error while reading " + csv.filename +
```
tu préfères faire ça plutot que faire un LOG4CPLUS_FATAL ?

antoine-de replied:
oui je pense que c'est mieux

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -262,124 +287,147 @@ void TripsFusioHandler::init(Data& d) {
 
-    auto vj_it = gtfs_data.vj_map.find(row[trip_c]);
-    if(vj_it != gtfs_data.vj_map.end()) {
+    if(gtfs_data.tz.vj_by_name.find(row[trip_c]) != gtfs_data.tz.vj_by_name.end()) {
```
Tu ignores ici ?

antoine-de replied:
oui si on trouve 2 fois le même on ne prend pas le second

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -262,124 +287,147 @@ void TripsFusioHandler::init(Data& d) {
+        std::string vj_uri = original_uri;
+        if (has_been_split) {
+            //we change the name of the vj (all but the first one) since we had to split the original GTFS vj because of dst
+            vj_uri += "_" + std::to_string(cpt_vj);
```
Dans l'idée se serait pas mal de vérifier que le nom n'existe pas déjà

antoine-de replied:
oui on va faire ca avec les meta vj

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -262,124 +287,147 @@ void TripsFusioHandler::init(Data& d) {
+
+        vj->validity_pattern = vp_xx;
+        vj->adapted_validity_pattern = vp_xx;
+        vj->journey_pattern = 0;
```
= nullptr non ?

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -262,124 +287,147 @@ void TripsFusioHandler::init(Data& d) {
+void TripsFusioHandler::handle_line(Data& data, const csv_row& row, bool is_first_line) {
+    auto split_vj = TripsFusioHandler::get_split_vj(data, row, is_first_line);
+
+    if (split_vj.empty())
```
des petits {} autour

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -262,124 +287,147 @@ void TripsFusioHandler::init(Data& d) {
+
+    for (auto vj: split_vj) {
+
+        if (is_valid(ext_code_c, row))
```
pareil

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -262,124 +287,147 @@ void TripsFusioHandler::init(Data& d) {
+            if (itm == gtfs_data.physical_mode_map.end()) {
+                LOG4CPLUS_WARN(logger, "TripsFusioHandler : Impossible to find the physical mode " << row[physical_mode_c]
+                               << " referenced by trip " << row[trip_c]);
+            }else{
```
les espaces

l-vincent-l commented on source/ed/connectors/gtfs_parser.cpp:
```diff
@@ -493,77 +673,95 @@ void TripsGtfsHandler::init(Data& data) {
+        std::string vj_uri = original_uri;
+        if (has_been_split) {
+            //we change the name of the vj (all but the first one) since we had to split the original GTFS vj because of dst
+            vj_uri += "_" + std::to_string(cpt_vj);
```
même remarque que précédemment sur l'existence du nom

antoine-de replied:
oui complétement

l-vincent-l commented on source/ed/connectors/gtfs_parser.h:
```diff
@@ -57,8 +84,8 @@ struct GtfsData {
-    std::unordered_map<std::string, ed::types::ValidityPattern*> vp_map;
-    std::unordered_map<std::string, ed::types::VehicleJourney*> vj_map;
+//    std::unordered_map<std::string, ed::types::ValidityPattern*> vp_map;
+//    std::unordered_map<std::string, ed::types::VehicleJourney*> vj_map;
```
tu peux les supprimer du coup

l-vincent-l commented on source/jormungandr/jormungandr/interfaces/v1/fields.py:
```diff
@@ -55,6 +58,59 @@ def __init__(self, *args, **kwargs):
+                if self.debug:
+                    logging.getLogger(__name__).debug('timezone is : {}'.format(cur))
+                tz = pytz.timezone(cur)
+                if not tz:
```
je pense que tu peux supprimer une indentation

l-vincent-l commented on source/jormungandr/jormungandr/interfaces/v1/fields.py:
```diff
@@ -55,6 +58,59 @@ def __init__(self, *args, **kwargs):
+
+        if value is None:
+            return self.default
+
```
Il faut aussi traiter le cas ou la zone n'est pas trouvée

l-vincent-l commented on source/jormungandr/jormungandr/interfaces/v1/fields.py:
```diff
@@ -320,7 +377,7 @@ def output(self, key, obj):
     "administrative_region": PbField(admin),
     "embedded_type": enum_type(),
     "name": fields.String(),
-    "quality" : fields.Integer(),
```
Je suis fier de toi

antoine-de committed:
Ed: corrections after review

antoine-de committed:
Ed: corrections after review

l-vincent-l commented on source/jormungandr/jormungandr/scripts/default.py:
```diff
@@ -539,17 +544,15 @@ def fill_journeys(self, pb_req, request, instance):
                     break
 
             new_datetime = None
+            one_minute = 1 * 60
```
= 60

kinnou02 commented on source/ed/connectors/gtfs_parser.cpp:
```diff
@@ -46,6 +46,114 @@ namespace ed{ namespace connectors {
+        if (tz) {
+            return {tz_name, tz};
+        }
+        LOG4CPLUS_INFO(log4cplus::Logger::getInstance("ed"), "cannot find " << tz_name << " in tz db");
```
maybe a warning

kinnou02 commented on source/ed/connectors/gtfs_parser.cpp:
```diff
@@ -46,6 +46,114 @@ namespace ed{ namespace connectors {
+std::vector<period_with_utc_shift> get_dst_periods(const boost::gregorian::date_period& validity_period, const boost::local_time::time_zone_ptr& tz) {
+
+    if (validity_period.is_null()) {
+        LOG4CPLUS_WARN(log4cplus::Logger::getInstance("log"), "validity period is not valid");
```
why not throw an exception?

antoine-de replied:
done

antoine-de committed:
Review: corrections after review

antoine-de committed:
Review: corrections after review

kinnou02 commented on source/ed/connectors/gtfs_parser.h:
```diff
@@ -47,6 +49,31 @@ www.navitia.io
+ */
+struct TzHandler {
+    TzHandler() {
+        tz_db.load_from_file(std::string(FIXTURES_DIR) + "date_time_zonespec.csv"); //TODO real file handling
```
it will not work in real life

antoine-de replied:
yup :)

kinnou02 commented on source/jormungandr/jormungandr/exceptions.py:
```diff
@@ -93,6 +93,13 @@ def __init__(self, arg):
+
+    def __init__(self, msg):
+        self.data = format_error("technical_error", msg)
+        self.code = 400
```
400 for a technical error?

antoine-de committed:
Jormungandr: change error code

kinnou02 commented on source/jormungandr/jormungandr/instance_manager.py:
```diff
@@ -140,24 +143,34 @@ def dispatch(self, arguments, api, instance_obj=None, instance_name=None,
                         instance.geom = None
+                else:
+                    instance.geom = None
+                instance.timezone = resp.metadatas.timezone
```
what do we do if it's None or worst, the field timezone doesn't exist?

antoine-de replied:
for null values I think it should be handled in the conversion.

For the field existence, we never check it (just above but I think it's historical and useless since we don't check the shape).

You could decide to always (or with all polling request) check the existence for compatibility reason, but that won't be the only thing to change

antoine-de commented:
test this please

antoine-de committed:
All: merge dev

antoine-de committed:
Ed: correction of the tests

antoine-de committed:
Ed: correction of the tests

antoine-de committed:
Ed: Add Meta VJ shell

a meta vj is a shell around some vj

antoine-de committed:
Ed: Add Meta VJ shell

a meta vj is a shell around some vj

antoine-de committed:
Ed: Fix associated calendar for meta vj

the association to calendar takes now the theoric vj of a meta vj, thus
we bypass the vj dst split

antoine-de committed:
Ed: corrections after review

antoine-de committed:
TimeTable: add (failing) tests on departure board

TDD power \o/

add a failing test on departure board, we now have to rework departure
board to make them OK

antoine-de committed:
TimeTable: refactor departure board for calendar

change the whole function to take into account the meta vj (because of
dst).

We now have to add time zone information

antoine-de committed:
Type: change init of vj fields

antoine-de committed:
All: rebase on dev

antoine-de committed:
TimeTable: fix departure board for calendar

In kraken all times are stored in UTC

however, for calendar we do not have a date to convert the time to a
local value (the conversion should be in jormungandr)

The for the calendar departure board api, we store in all vj the utc
offset needed to convert all stop times of the vehicle journey to local.

Thus:
* when queried with a date time the departure board answers in UTC and
the conversion is done in jormun
* when queried with a calendar the departure board answers in local
time

antoine-de committed:
Protobuff: change date from int32 to int64

protobuff compact the types and google recommendation for dates are
uint64, so we're safe after 2038 :p

antoine-de committed:
Jormun: add test

l-vincent-l committed:
merge fixes

l-vincent-l committed:
Ed: fix a merge error of PR 471

l-vincent-l committed:
Jormungandr: Delete double car direct_path

l-vincent-l committed:
Jormungandr: display stop_points of a route

antoine-de committed:
GTFS: ODT flag on stoptime

thanks @skywave for the issue, the stop time is odt if either it is an
odt drop of or an odt pickup

antoine-de committed:
Cmake: add full path to xunit ouput

the new cmake version behave differently than the previous one, so add
the full path to the output

antoine-de committed:
All: change datetime conversion to int 64

thanks @skywave for the review again

antoine-de committed:
Gtfs: hard code boost timezone csv

Uggly but necessary wrapper for the moment

Unfortunately the tz db is not supported by boost::date_time (it is done
with boost::locale but we don't want to pay the cost of this module).

boost::date_time can only be fed with a csv or with code.
We don't want to bother the delivery of the file with the binary so we
are stuck with the hardcoding of the timezone data

The clean way would be to change boost to read the tz db and we'll see
if we can manage some time to do that in the future.

antoine-de commented:
a bit of  a mess up because a nonwrite right on other's repository.
I rebased @l-vincent-l 's branch (who rebased from mine :p). So I close the PR #493 and the only timezone PR for the timezone refactoring is this one.

kinnou02 commented:
ok for me

koch-t commented:
How sure are you that non on your transit-operators add trips on nights with DST switch without explicitly indicating that they will never ride due to the DST switch?
For example at 30th of March at least two operators in the Netherlands included trips with departure-times of:

00:00:00
01:00:00
02:00:00 (Never rides)
03:00:00 

The undefined behavior they expect is that the planner will never show the 02:00 trip because 02:00 will be the 03:00 trip

antoine-de commented:
I think most french operator also give the data in local datetime (and not split by dst period) without additional information.

In the current implementation we'll indeed have either a 'hole' in the schedule or a dupplicate, but I don't really see how we can do better without additional information.

antoine-de committed:
datetime: delete csv

delete boost timezone csv since it is now handled in code

antoine-de committed:
Utils: revert submodule after rebase
